play/src/pusher/models/Space.ts

```diff
@@ -0,0 +1,345 @@
+                // I'm sure user uuid is existing
+                user_ = this.users.get(subMessage.message.updateSpaceUserMessage.user?.uuid as string);
+            }
+            if (user_) {
+                return this.filterOneUser(spaceFilter, user_);
+            }
+        }
+        return false;
+    }
+
+    public filter(spaceFilter: SpaceFilterMessage) {
+        return [...this.users.values()].filter((user) => this.filterOneUser(spaceFilter, user));
+    }
+
+    private filterOneUser(spaceFilters: SpaceFilterMessage, user: SpaceUser) {
+        if (spaceFilters.filter?.$case === "spaceFilterContainName") {
+            const spaceFilterContainName = spaceFilters.filter.spaceFilterContainName;
+            if (user.name.includes(spaceFilterContainName.value)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public handleAddFilter(watcher: ExSocketInterface, updateSpaceFilterMessage: UpdateSpaceFilterMessage) {
+        const newFilter = updateSpaceFilterMessage.spaceFilterMessage;
+        if (newFilter) {
+            debug(`${this.name} : filter added (${newFilter.filterName}) for ${watcher.userUuid}`);
+            const oldData = [...this.users.values()];
+            const newData = this.filter(newFilter);
+            this.delta(watcher, oldData, newData, newFilter.filterName);
+        }
+    }
+
+    public handleUpdateFilter(watcher: ExSocketInterface, updateSpaceFilterMessage: UpdateSpaceFilterMessage) {
+        const newFilter = updateSpaceFilterMessage.spaceFilterMessage;
+        if (newFilter) {
+            const oldFilter = watcher.spacesFilters.find((filter) => filter.filterName === newFilter.filterName);
+            if (oldFilter) {
+                debug(`${this.name} : filter updated (${newFilter.filterName}) for ${watcher.userUuid}`);
+                const oldData = this.filter(oldFilter);
+                const newData = this.filter(newFilter);
+                this.delta(watcher, oldData, newData, newFilter.filterName);
+            }
+        }
+    }
+
+    public handleRemoveFilter(watcher: ExSocketInterface, removeSpaceFilterMessage: RemoveSpaceFilterMessage) {
+        const oldFilter = removeSpaceFilterMessage.spaceFilterMessage;
+        if (oldFilter) {
+            debug(`${this.name} : filter removed (${oldFilter.filterName}) for ${watcher.userUuid}`);
+            const oldData = this.filter(oldFilter);
+            const newData = [...this.users.values()];
+            this.delta(watcher, oldData, newData, undefined);
+        }
+    }
+
+    private delta(
+        watcher: ExSocketInterface,
+        oldData: SpaceUser[],
+        newData: SpaceUser[],
+        filterName: string | undefined
+    ) {
+        // Check delta between responses by old and new filter
+        const addedUsers = newData.filter(
+            (leftValue) => !oldData.some((rightValue) => isSameUser(leftValue, rightValue))
+        );
```

Comment:
The performance of this is `O(n^2)` (one loop for `filter` and one loop for `some`)
We must find a more efficient way to compare the 2 lists.

I think we can do so by using Maps or Sets instead of lists.

The filter() method should probably return a Map instead of an Array.